File: interview-keyboard-challenge-master/obj-c/InterviewChallengeUtilities.m
//
//  InterviewChallengeUtilities.m
//  interview-keyboard-challenge
//

#import "InterviewChallengeUtilities.h"

@implementation InterviewChallengeUtilities
+ (NSDictionary *)getUsers:(NSString *)filePath{
    NSData *data = [NSData dataWithContentsOfFile:filePath];
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
    NSDictionary *users = json[@"users"];
    
    NSMutableDictionary *usersDictionary = [NSMutableDictionary dictionaryWithCapacity:users.count];
    [users enumerateKeysAndObjectsUsingBlock:^(NSString *user_id, NSDictionary *userDictionary, BOOL *stop) {
        usersDictionary[user_id] = [User userWithDictionary:userDictionary];
    }];
    
    return [usersDictionary copy];
}

+ (NSArray *)getEvents:(NSString *)filePath{
    NSData *data = [NSData dataWithContentsOfFile:filePath];
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
    NSArray *events = json[@"events"];
    
    NSMutableArray *eventsArray = [NSMutableArray arrayWithCapacity:events.count];
    [events enumerateObjectsUsingBlock:^(NSDictionary *eventDictionary, NSUInteger idx, BOOL *stop) {
        [eventsArray addObject:[Event eventWithDictionary:eventDictionary]];
    }];
    
    return [eventsArray copy];
}

@end

@implementation User
- (instancetype)initWithDictionary:(NSDictionary *)jsonDictionary
{
    self = [super init];
    if (self) {
        _age = [jsonDictionary[@"age"] integerValue];
        _gender = jsonDictionary[@"gender"];
        _device = jsonDictionary[@"device"];
    }
    return self;
}

+ (instancetype)userWithDictionary:(NSDictionary *)jsonDictionary
{
    return [[self alloc]initWithDictionary:jsonDictionary];
}
@end

@implementation Event
- (instancetype)initWithDictionary:(NSDictionary *)jsonDictionary
{
    self = [super init];
    if (self) {
        _name = jsonDictionary[@"name"];
        _timestamp = [jsonDictionary[@"timestamp"] integerValue];
        _user_id = jsonDictionary[@"user_id"];
    }
    return self;
}

+ (instancetype)eventWithDictionary:(NSDictionary *)jsonDictionary
{
    return [[self alloc]initWithDictionary:jsonDictionary];
}
@end

File: interview-keyboard-challenge-master/swift/InterviewChallengeUtilities.swift
//
//  InterviewChallengeUtilities.swift
//

import Foundation

public struct User{
    public var age:Int
    public var gender:String
    public var device:String
    
    init(json:Dictionary<String,AnyObject>){
        age = json["age"] as! Int
        gender = json["gender"] as! String
        device = json["device"] as! String
    }
}

public struct Event{
    public var name: String
    public var timestamp: Int
    public var user_id: String

    init(json:Dictionary<String,AnyObject>){
        name = json["name"] as! String
        timestamp = json["timestamp"] as! Int
        user_id = json["user_id"] as! String
    }
}

typealias Users = Dictionary<String, User>
typealias Events = Array<Event>

func getUsers(filePath: String) -> Users {
    var users = Users()

    do {
        
        let data = try NSData(contentsOfFile: filePath) as Data
        
        if let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] {
            
            let usersJSON = json["users"] as! Dictionary<String,Dictionary<String,AnyObject>>
            for (user_id,userDictionary) in usersJSON{
                users[user_id] = User(json: userDictionary)
            }
        }
    }
    catch let error {
        print("Failed to read users from disk. Error: \(error.localizedDescription)")
    }
    
    return users
}

func getEvents(filePath:String) -> Events{
    var events = Events()
    
    do {
        
        let data = try NSData(contentsOfFile: filePath) as Data
        
        if let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any] {
            
            let eventJSON = json["events"] as! Array<Dictionary<String,AnyObject>>
            for eventDictionary in eventJSON{
                events.append(Event(json: eventDictionary))
            }
        }
    }
    catch let error {
        print("Failed to read events from disk. Error: \(error.localizedDescription)")
    }

    return events
}